Add a `test` implementation with `shunit2`

Eric-Arellano opened the pull request:
Corresponds to https://www.pantsbuild.org/v2.0/docs/plugins-test-goal (WIP).

https://github.com/kward/shunit2 is a neat script to run tests with Bash. It simply requires having the script as a sibling file to your test file, then adding `source ./shunit2` at the bottom. Pants makes this even more convenient by ensuring that the script is always present as a sibling to your test file and by automatically adding `soruce ./shunit2` to the bottom of the file if it's not already present.

We add a `bash_setup.py` file so that users can set the `$PATH`, and so that we can factor out common code to find the `bash` program.

This PR also improves the documentation for the rest of this repo.

Eric-Arellano committed:
Add a `test` implementation with `shunit2`

cosmicexplorer commented on pants-plugins/examples/bash/bash_setup.py:
```diff
@@ -0,0 +1,85 @@
+    @classmethod
+    def register_options(cls, register):
+        register(
+            "--executable-search-paths",
```
This is great!! I like that we can expose just one part of the environment PATH and not just take the parent env!

Eric-Arellano replied:
Credit to John for the idea! We have an equivalent option `--pex-executable-search-paths`.

cosmicexplorer commented on pants-plugins/examples/bash/bash_setup.py:
```diff
@@ -0,0 +1,85 @@
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-installing-tools.
+    bash_program_paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(binary_name="bash", search_path=["/bin", "/usr/bin"]),
```
This is so, so nice.

cosmicexplorer replied:
Is this `search_path` intended to be set to the value of the `executable_search_path` property after `<PATH>` expansion? I would have assumed that was what that option was for.

Eric-Arellano replied:
Yes! Credit to John. This is one of my favorite recent additions to the Plugin API. See https://www.pantsbuild.org/v2.0/docs/rules-api-installing-tools if you haven't yet. I remember when working on a Docker plugin with a user in 1.25, the hardest question was "How do I access Docker..?" We now have a couple good answers for that :)

> this search_path intended to be set to the value of the executable_search_path property 

Oh, hm, yeah, it probably should be. Thanks.

cosmicexplorer replied:
It certainly automates a lot of what I was previously doing in `XCodeCLITools` in the old native backend, that will be much easier to write safely now (it had several path searching bugs that had to be fixed).

cosmicexplorer commented on pants-plugins/examples/bash/lint/shfmt/rules.py:
```diff
@@ -60,6 +66,7 @@ async def setup_shfmt(setup_request: SetupRequest, shfmt: Shfmt) -> Setup:
 
     # If the user specified `--shfmt-config`, we must search for the file they specified with
     # `PathGlobs` to include it in the `input_digest`. We error if the file cannot be found.
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
```
```suggestion
    # See https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
```

Eric-Arellano replied:
Ah, thanks!

cosmicexplorer commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+    shunit2_script_request = Get(
+        Digest,
+        DownloadFile(
+            url="https://raw.githubusercontent.com/kward/shunit2/b9102bb763cc603b3115ed30a5648bf950548097/shunit2",
```
Oh if we're using the raw github blobs, this could totally be versioned. I would at least leave a TODO to use an `ExternalTool` here if not directly in this PR.

Eric-Arellano replied:
Counterpoint is that there's a tension with example repos between making things really robust vs. keeping them simple so that they're easier to follow. While `ExternalTool` would make this more comprehensive, I fear it would complicate understanding the example. And I think it's valuable to demo `DownloadFile`, as this is the first time we do that. Thoughts?

cosmicexplorer replied:
Ok, I totally agree with that. Could we at least put a link to the `ExternalTool` docs at https://www.pantsbuild.org/v2.0/docs/rules-api-installing-tools#externaltool-install-pre-compiled-binaries and note that's a more robust solution that lets users set the version?

Eric-Arellano replied:
Yes, we do that right above. 

```python
    # Because shunit2 is a simple Bash file, we download it using `DownloadFile`. Normally, we
    # would install the test runner through `ExternalTool`. See
    # https://www.pantsbuild.org/v2.0/docs/rules-api-installing-tools and
    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
```

cosmicexplorer replied:
oh, lmao. thanks!!

cosmicexplorer commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+    # https://www.pantsbuild.org/v2.0/docs/rules-api-file-system.
+    #
+    # Most test runners don't modify their test files like we do here, so most test runners can
+    # skip this step.
```
Do we need to check this every time an `shunit2` test runs? Or can we move this check into a test file that only runs in our CI?

Eric-Arellano replied:
Yes. If the file doesn't already have `source ./shunit2`, then no tests will run for it. It doesn't matter what environment we're running the tests in. So long as it's a `bash_tests` file, we need to guarantee that `source ./shunit2` is present.

This is all an affordance that Pants is making, though. We could instead make the user manually include that line.

cosmicexplorer replied:
Oh, sorry. You're totally right.

cosmicexplorer commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+@rule(desc="Setup Shunit2 to run interactively", level=LogLevel.DEBUG)
+async def setup_shunit2_debug_test(field_set: Shunit2FieldSet) -> TestDebugRequest:
+    setup = await Get(TestSetup, TestSetupRequest(field_set))
+    # We set up an InteractiveProcess, which will get run in the `@goal_rule` in `test.py`.
```
```suggestion
    # We set up an InteractiveProcess, which will be executed in the `@goal_rule` in `test.py`.
```
*extremely* subjective change, feel free to ignore

Eric-Arellano replied:
I like it. Thanks.

cosmicexplorer reviewed (commented):
I have a question about `PATH` searching for the `bash` and `zip` executables, but otherwise this is fantastic and extremely well-documented!!!

Eric-Arellano committed:
Review feedback

Thanks Danny!

Eric-Arellano committed:
Fix fmt

Eric-Arellano committed:
Fix and improve error messages for `zip` and `bash` not being found

Originally I was going to factor this code up in Pants, but I realized that that would result in worse error messages. We should keep the error messages local so that we can give the specific remedy for fixing the issue, which varies depending on how the plugin is implemented.

benjyw commented on pants-plugins/examples/bash/lint/shellcheck/register.py:
```diff
@@ -1,9 +1,14 @@
+
+Note that we use a separate `register.py` than the main `examples/bash/register.py` so that users
+must opt into activating Shellcheck. We could have instead registered the Shellcheck rules there
+if we were okay with Shellcheck being activated when the Bash backend is activated.
```
```suggestion
if we were okay with Shellcheck always being activated when the Bash backend is activated.
```

benjyw commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+After setting up the source files (including finding all relevant transitive dependencies), we run
+the equivalent of `bash test_file.sh`.
+
+We must implement rules for both normal `test` and `test --debug`. To deduplicate, we have a
```
```suggestion
We must implement rules for both normal `test` and `test --debug`. To avoid duplication, we have a
```

benjyw commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+This plugin uses the Shunit2 test runner script (https://github.com/kward/shunit2). Because the
+test runner is a simple Bash script, we simply download the file to be able to run it. Shunit2
+requires that tests have "source ./shunit2" at the bottom of the tests; as a convenience, our
+plugin uses file system operations to automatically add that line to any test files missing it.
```
```suggestion
plugin uses filesystem operations to automatically add that line to any test files missing it.
```

benjyw commented on pants-plugins/examples/bash/shunit2_test_runner.py:
```diff
@@ -0,0 +1,181 @@
+        dependencies_source_files_request, test_source_files_request
+    )
+
+    # To check if the test files already have `source ./shunit2` in them, we need to look at the
```
TBH I'm not sure this is necessary, and it complicates the example. We're supposed to be "a thin wrapper around tools", and "we will edit your source files behind your back" is very not-that.  

I think if someone is writing a shunit2 test, and must already be aware of how shunit2 works, then expecting them to add necessary boilerplate is totally fine, rather then being too-clever-by-half.

Eric-Arellano replied:
Hm, would you expect us to still download shunit2 automatically?

Would the user need to put `source ./shunit2` (sibling file), or we should attempt to understand `source ../../shunit2`?

benjyw replied:
Ah, I see, that does muddy the waters. 

What are users expected to do in general, check shunit2 into their repos?  That would, for example, guarantee a specific version.

Eric-Arellano replied:
Yeah, I think so. And then reference it relative to where it's checked in.

But this is a huge selling point of Pants: we will get the tools for you, whether it be shunit2 or Pytest. That's really nice imo that now you can write a test anywhere in your repo without having to remember where you checked in the `./shunit2` script.

benjyw replied:
Fair. And we don't prevent people from doing the manual thing if they want.

Eric-Arellano replied:
Exactly

Eric-Arellano committed:
Apply suggestions from code review

Co-authored-by: Benjy Weinberger <benjy@toolchain.com>
